server/service/redis_policy_set/redis_policy_set.go

```diff
@@ -0,0 +1,167 @@
+// Package redis_policy_set provides a Redis implementation of service.FailingPolicySet.
+package redis_policy_set
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/fleetdm/fleet/v4/server/datastore/redis"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	redigo "github.com/gomodule/redigo/redis"
+)
+
+type redisFailingPolicySet struct {
+	pool fleet.RedisPool
+}
+
+var _ fleet.FailingPolicySet = (*redisFailingPolicySet)(nil)
+
+// NewFailing creates a redis policy set for failing policies.
+func NewFailing(pool fleet.RedisPool) *redisFailingPolicySet {
+	return &redisFailingPolicySet{
+		pool: pool,
+	}
+}
+
+const (
+	policySetKeyPrefix = "policies:failing:"
+	// policySetsSetKey is used to avoid a SCAN command when listing policy sets.
+	policySetsSetKey = "policies:failing_sets"
+)
+
+// ListSets lists all the policy sets.
+func (r *redisFailingPolicySet) ListSets() ([]uint, error) {
+	conn := redis.ConfigureDoer(r.pool, r.pool.Get())
+	defer conn.Close()
+
+	ids, err := redigo.Ints(conn.Do("SMEMBERS", policySetsSetKey))
+	if err != nil && err != redigo.ErrNil {
+		return nil, err
+	}
+	policyIDs := make([]uint, len(ids))
+	for i := range ids {
+		policyIDs[i] = uint(ids[i])
+	}
+	return policyIDs, nil
+}
+
+// AddHost adds the given host to the policy sets.
+func (r *redisFailingPolicySet) AddHost(policyID uint, host fleet.PolicySetHost) error {
+	conn := redis.ConfigureDoer(r.pool, r.pool.Get())
+	defer conn.Close()
+
+	// The order of the following two operations is important.
+	//
+	// The ordering of operations in AddHost and RemoveSet has been chosen to avoid
+	// ending up with a policySetKey with a host entry, but without its corresponding entry
+	// in the set of sets.
+	if _, err := conn.Do("SADD", policySetKey(policyID), hostEntry(host)); err != nil {
+		return err
+	}
+	if _, err := conn.Do("SADD", policySetsSetKey, policyID); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *redisFailingPolicySet) scanPolicySet(conn redigo.Conn, policyID uint, count int) ([]string, error) {
+	var hosts []string
+
+	cursor := 0
+	for {
+		res, err := redigo.Values(conn.Do("SSCAN", policySetKey(policyID), cursor, "COUNT", count))
```

Comment:
What's the reasoning about COUNT being always zero?

Also:

```suggestion
		res, err := redigo.Values(conn.Do("SSCAN", policySetKey(policyID), cursor, "COUNT", 0))
```

Reply:
Maybe you misread cursor vs count?

Reply:
Ah, pfft... yeah, ignore!